### PR TITLE
[IMP] wowl: CrashManager

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
+import { Component, tags, hooks } from "@odoo/owl";
+import { DialogParams, OdooEnv } from "../../types";
+import { Dialog } from "../dialog/dialog";
+import { RPCError, RPCServerError } from "../../services/rpc";
+import { Stringifiable, _lt } from "../../core/localization";
+
+const { useState } = hooks;
+
+const odooExceptionTitleMap: { [name: string]: Stringifiable } = {
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts]: what about defining this inside the crash manager? That way you could use env._t

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+  }
+
+  _onRPCError(error: RPCError) {
+    // CASE 1: not a server error
+    if (error.type !== "server") {
+      return;
+    }
+
+    // CASE 2: lost connection
+    if (error.code === -32098) {
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts]: lol

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      // CASE 4: Odoo custom exception
+      const isOdooError = !!odooExceptionTitleMap[error.data_name];
+      if (isOdooError) {
+        const message = error.message;
+        const title = odooExceptionTitleMap[error.data_name].toString();
+        this._addDialog({ message, title });
+      }
+    }
+
+    // CASE 5: code generic for arbitrary python error
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts]: generic code

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    /** to do */
+  }
+
+  _onDialogClosed(id: number) {
+    delete this.crashes[id];
+  }
+
+  _onButtonClick(id: number, button: { text: string | Stringifiable; onClick?: Function }) {
+    if (button.onClick) {
+      button.onClick(); // bind to this? add parameters?
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts]: is this for case 3 only?

File: addons/wowl/static/src/types.ts [hunk 5]
```diff
@@ -158,6 +157,14 @@ export interface SystrayItem {
   sequence?: number;
 }
 
+export interface DialogParams {
+  message: string | Stringifiable;
+  title: string | Stringifiable;
+  buttons?: { text: string | Stringifiable; onClick?: Function }[];
```
Comment (aab-odoo) [inline addons/wowl/static/src/types.ts]: for onClick, you know the args

**Review by aab-odoo — COMMENTED** [review]
this is looking good. Not sure about the buttons stuff either though

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 6]
```diff
@@ -0,0 +1,355 @@
...
+    line: number,
+    col: number,
+    error: any /** Object with possibly stack */
+  ) {
+    // Scripts injected in DOM (eg: google API's js files) won't return a clean error on window.onerror.
+    // The browser will just give you a 'Script error.' as message and nothing else for security issue.
+    // To enable onerror to work properly with CORS file, you should:
+    //   1. add crossorigin="anonymous" to your <script> tag loading the file
+    //   2. enabling 'Access-Control-Allow-Origin' on the server serving the file.
+    // Since in some case it wont be possible to to this, this handle should have the possibility to be
```
Comment (kebeclibre) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts]: toutou :)

Comment (Polymorphe57) [reply addons/wowl/static/src/components/crash_manager/crash_manager.ts]: ok :-)

File: addons/wowl/static/src/components/crash_manager/crash_manager.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+  isConnected: boolean = true;
+  crashes = useState({} as Crashes);
+  crashId: number = 1;
+  notificationService = useService("notifications");
+  rpc = useService("rpc");
+
+  constructor() {
+    super(...arguments);
+    useExternalListener(window, "error", this._onError);
+    /** @todo: Before, this event was handled first by crash manager then by QUnit --> should we do the same: negate some erros? */
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/crash_manager/crash_manager.ts:45]: errors

File: addons/wowl/static/src/components/dialog/dialog.xml [hunk 8]
```diff
@@ -13,7 +13,7 @@
                         <div class="modal-content">
                             <header t-if="props.renderHeader" class="modal-header">
                                 <h4 class="modal-title">
-                                    <t t-esc="props.title"/>
+                                    <t t-esc="env._t(props.title)"/>
```
Comment (aab-odoo) [inline addons/wowl/static/src/components/dialog/dialog.xml]: does it work?

Comment (Polymorphe57) [reply addons/wowl/static/src/components/dialog/dialog.xml:16]: yes

File: addons/wowl/doc/crash_manager.md [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+The `crash manager` is a component whose responsibility is to display a dialog
+when an rpc error or a client error occurs. It is one of the webclient main
+components (see [Main Components](main_components.md)).
+
+## Channels
+
+The crash manager receives errors throught two channels:
+
+- it listens on `env.bus` the event `RPC_ERROR`;
+- it listens on `window` the event `error` throught the use of the hook `useExternalListener`;
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: useExternalListener is an implementation detail imho

File: addons/wowl/doc/crash_manager.md [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+## Channels
+
+The crash manager receives errors throught two channels:
+
+- it listens on `env.bus` the event `RPC_ERROR`;
+- it listens on `window` the event `error` throught the use of the hook `useExternalListener`; 
+
+## RPC_ERROR event handling
+
+When an event `RPC_ERROR` is triggerd on the bus, the crash manager will processe the
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: process

Comment (aab-odoo) [reply addons/wowl/doc/crash_manager.md]: actually, I would write "the crash manager processes the..."

File: addons/wowl/doc/crash_manager.md [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+`RPCError` in the following way:
+
+- look if the error's `type` is `server`;
+
+- if this is the case, the optional error's `name` indicates which
+  dialog class should be used to display the error details.
+  If the error is unammed or no dialog class corresponds to its name, the class
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: unnamed

File: addons/wowl/doc/crash_manager.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+  `ErrorDialog` is used by default.
+
+- The dialog class is instantiated with a unique prop the error itself.
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: unique prop: the error itself (?)

File: addons/wowl/doc/crash_manager.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+
+This how the `UserError`, `AccessError`, ... or the custom server errors are handled.
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: typography rule: no `,` before `...`

File: addons/wowl/doc/crash_manager.md [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+
+## ERROR event handlling
+
+When an event `error` is triggered on window, the crash manager will process the `ErrorEvent`
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: will process -> processes

File: addons/wowl/doc/crash_manager.md [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+received in the following way:
+
+- if some information on the file name where the errors occurs, the error stack,... is available
```
Comment (aab-odoo) [inline addons/wowl/doc/crash_manager.md]: error

Comment (aab-odoo) [reply addons/wowl/doc/crash_manager.md]: occurred

File: addons/wowl/doc/registries.md [hunk 16]
```diff
@@ -7,8 +7,11 @@ the web client in a safe and structured way.
...
-| Name         | Description                                                        |
-| ------------ | ------------------------------------------------------------------ |
-| `Components` | components (class) that will be instantiated at root of web client |
-| `services`   | definition of all services that will be deployed                   |
-| `views`      | definition of all available views                                  |
+| Name         | Description                                                                         |
+| ------------ | ----------------------------------------------------------------------------------- |
+| `actions`    | definition of all available client actions                                          |
+| `Components` | components (class) that will be instantiated at root of web client                  |
+| `dialogs`    | dialogs (class) that will be instantiated by the crash manager to handle errors     |
```
Comment (aab-odoo) [inline addons/wowl/doc/registries.md]: errorDialogs? I'm afraid 'dialogs' is too generic

File: addons/wowl/doc/services/rpc.md [hunk 17]
```diff
@@ -71,11 +71,18 @@ When a rpc fails, then:
...
+
+  - `name(string)` (optional, used by the crash manager to look for an appropriate
+                    dialog to use when handling the error)
+  - `subType(string)` (optional, often used to determine the dialog title)
+  - `data(object)` (optional object that can contain various keys among which
+                    `debug`: the main debug information, with the call stack)
 
   If it is a network error, then the error description is simply an object
   `{type: 'network'}`.
+  When a network error occurs, a notification will be displayed and the server will be
```
Comment (aab-odoo) [inline addons/wowl/doc/services/rpc.md]: is displayed

**Pull request merged**